src/internal.rs

```diff
@@ -242,7 +361,16 @@ pub trait Parser<Input> {
 
   /// A parser takes in input type, and returns a `Result` containing
   /// either the remaining input and the output value, or an error
-  fn parse(&mut self, input: Input) -> IResult<Input, Self::Output, Self::Error>;
+  fn parse(&mut self, input: Input) -> IResult<Input, Self::Output, Self::Error> {
+    self.process::<OutputM<Emit, Emit, Streaming>>(input)
+  }
+
+  /// A parser takes in input type, and returns a `Result` containing
+  /// either the remaining input and the output value, or an error
+  fn process<OM: OutputMode>(
+    &mut self,
+    input: Input,
+  ) -> PResult<OM, Input, Self::Output, Self::Error>;
```

Comment:
> test the effect on benchmarks

I'm curious to see if you are able to see the dramatic savings that chumsky did or if some of that is more inherent to their design.  My gut feeling is that in a lot of cases, the overhead from a lot of these outputs and error should be pretty low with nom's design as-is and that this wouldn't offer much improvement unless people were going heavy on complex `preceded`  / `termianted` parsers which json doesn't do and I've not done in my various parsers.

Besides runtime performance, it might also be good to check the impact to binary size or release binaries as this might cause more instances of the generic functions to be generated.  I'd also be curious about compile times but I'm assuming that won't be big enough to notice.

Reply:
currently most of nom's overhead is in creating values and errors and discarding them, and that will be even more present with heavy error types. So far I have not measured it though, because I have not converted enough code to affect the benchmarks. I'm worried about compilation times, though. That might be a mix of increasing compilation times due to heavy use of impl traits (returning closures was erasing type information), but decreasing it due to less code  being generated.

Reply:
> currently most of nom's overhead is in creating values and errors and discarding them

Can you qualify this with some more details?

For example, if you are doing `delimited(char('['), <something>, char(']'))`, then I would expect the value wouldn't have much overhead.  From the nom examples I've looked at and in my own parsers, this is a fairly common case.  Is the overhead more than I expect or is this for more complex usage patterns?  chumsky saw the gains on `json` but I see these more simple cases in nom's json benchmark.

One possible gain is people not discover the `_count` variant of `many` parsers.  I know I make that mistake.  However, that also ties into needing to also pay attention to using `terminated` / `preceded` rather than tuples     Something I was considering was an  alternative trait to `Extend` that (1) allowed pre-allocation and (2) was implemented for `usize` and `()`, replacing `_count` variants.  This is a bottom-up driven approach to not creating expensive values, rather than the Mode's top-down approach.  One benefit to top-down is when reusing parsers between throw-away and non-throwaway contexts.  So far, I've not needed to do it too much and I would be interested in looking at cases for those that do and how much it impacts performance.

However, I can see it in errors with `alt`.  With `combine`, I did see performance overhead from their version of `alt` with heavy errors.  On the other hand, I suspect it'd be better to replace it with combine's  [dispatch](https://docs.rs/combine/latest/combine/macro.dispatch.html).  It is a macro that lets you write a `match` with parsers which cuts down on the number of parse attempts made, reducing overhead further.  I can only speak for switching `choice` / `alt` to `dispatch` but it  dramatically sped up my code.  While this can be done with an imperative parser in `nom`, I found [duplicating `dispatch`](https://github.com/toml-rs/toml/blob/main/crates/toml_edit/src/parser/macros.rs) made it more ergonomic which makes it more likely to be used.   I suspect some of my speed gains over `combine` were from finding another place I could insert a `dispatch` call.

Even with that, I've noticed my performance is sensitive to changes in my error type and I've intentionally kept  it simple, putting all of the expensive operations when converting to the public error type for my crate.  An `opt` that doesn't apply can still easily do a single allocation that is thrown away.  Dropping that allocation and dropping any of the panic handling around allocations can be beneficial.  I was considering a non-GAT approach here as well by making the backtracking error type accepted by `opt` to be `()`).  GATs do simplify the process.